File: python/algo/202606/algo3635.py
from cmath import inf
from typing import List

# 贪心 八、其他
class Solution:
    def earliestFinishTime(self, landStartTime: List[int], landDuration: List[int], waterStartTime: List[int], waterDuration: List[int]) -> int:
        # land -> water
        mn_fin_land = inf
        for s, d in zip(landStartTime, landDuration):
            mn_fin_land = min(mn_fin_land, s + d)
        
        res = inf
        for s, d in zip(waterStartTime, waterDuration):
            res = min(res, max(mn_fin_land, s) + d)

        # water -> land
        mn_fin_water = inf
        for s, d in zip(waterStartTime, waterDuration):
            mn_fin_water = min(mn_fin_water, s + d)
        
        for s, d in zip(landStartTime, landDuration):
            res = min(res, max(mn_fin_water, s) + d)

        return res

File: cplus/leetcode/202606/algo3635.cpp
#include <bits/stdc++.h>
using namespace std;

// 贪心 八、其他
class Solution {
public:
    int earliestFinishTime(vector<int>& landStartTime, vector<int>& landDuration, vector<int>& waterStartTime, vector<int>& waterDuration) {
        // 自行解答，变量太复杂，容易写错
        int ans = INT_MAX, mn_fin_land = INT_MAX, mn_fin_water = INT_MAX;
        int m = landStartTime.size(), n = waterStartTime.size();

        for (int i=0; i<m; i++) {
            auto s = landStartTime[i], d = landDuration[i];
            mn_fin_land = min(mn_fin_land, s + d);
        }

        for (int i=0; i<n; i++) {
            auto s = waterStartTime[i], d = waterDuration[i];
            ans = min(ans, max(mn_fin_land, s) + d);
        }

        for (int i=0; i<n; i++) {
            auto s = waterStartTime[i], d = waterDuration[i];
            mn_fin_water = min(mn_fin_land, s + d);
        }

        for (int i=0; i<m; i++) {
            auto s = landStartTime[i], d = landDuration[i];
            ans = min(ans, max(mn_fin_water, s) + d);
        }

        return ans;
    }

};